### DISCUSSION TITLE
What is the standard way to fetch deeply nested relation objects?

### DISCUSSION THREAD

**Discussion by xoldd** [urql-graphql/urql#2949] [Q&A]
@kitten @JoviDeCroock What is the standard way to fetch deeply nested relation objects? I'm listing a few methods that I thought of on the top of my head. Methods 1 and 2 make useless network calls to first fetch parent object then the relation objects. Method 3 is anti-pattern. Am I missing some holy grail of graphql data fetching mechanism here?

Fetch deeply nested objects in one query :- 
```
// Or equivalent relay node interface approach
query User($after_posts: String, $after_comments: String, $first: Int = 20, $id: String!){
    user(id: $id){
        postsConnection(after: $after_posts, first: $first){
            edges{
                node{
                    commentsConnection(after: $after_comments, first: $first){
                        edges{
                            node{
                                id
                            }
                        }
                    }
                }
            }
        }
    }
}
```

Fetch relation objects using parent object's id :-
```
// Or equivalent relay node interface approach
query Post($after_comments: String, $first: Int = 20, $id: String!){
    post(id: $id){
        commentsConnection(after: $after_comments, first: $first){
            edges{
                node{
                   id
                }
            }
        }
    }
}
```

Fetch relation objects using seperate query field(Anti graphQL pattern) :-
```
query CommentsConnectionByPost($after: String, $first: Int = 20, $postId: String!){
    commentsConnectionByPost(after: $after, first: $first, postId: $postId){
        edges{
            node{
                id
            }
        }
    }
}
```

**Comment by JoviDeCroock** [comment]
When I approach issues like these I like to rewind to considering your graph like a set of entry-points to your front-end. Let's say we have a posts-comments-user application like you are showing here.

When we enter the app anonymously we would enter from i.e.

```graphql
query {
  posts {
    edges { node { id comments { totalCount } author { id } }
  }
}
```

When we go into one of these posts it might be a post-by-id query but that's about the gist of it 😅 we try to get the totality of data for our entry-point. We hit the server once for all the data in a view.

Method 1 and 2 don't make "useless network calls" as it's only 1 call going to your back-end, you might need to optimize the way you are retrieving data with i.e. dataloader or having something like pothos who does that for you when using prisma.

**Reply by xoldd** [reply]
> When I approach issues like these I like to rewind to considering your graph like a set of entry-points to your front-end. Let's say we have a posts-comments-user application like you are showing here.
> 
> When we enter the app anonymously we would enter from i.e.
> 
> ```graphql
> query {
>   posts {
>     edges { node { id comments { totalCount } author { id } }
>   }
> }
> ```
> 
> When we go into one of these posts it might be a post-by-id query but that's about the gist of it sweat_smile we try to get the totality of data for our entry-point. We hit the server once for all the data in a view.
> 
> Method 1 and 2 don't make "useless network calls" as it's only 1 call going to your back-end, you might need to optimize the way you are retrieving data with i.e. dataloader or having something like pothos who does that for you when using prisma.

I understand the concept of fetching everything required on a page in one request. Then passing the fetched data exactly in the shape the child components asked for it using fragments. I'm talking about nested connections here.

Method 1 and 2 make useless network calls in the context of nested `commentsConnection` because pagination requires running the query multiple times with different variables. Method 2 for example would first need to resolve the parent `Post` object and only after that it would resolve the nested `commentsConnection` object. 

If we used method 3 we'd save on one network request by resolving `commentsConnection` directly by querying the database for comments where foreign key `postId` is equal to argument `postId`.

**Reply by kitten** [reply]
I think you're taking the idea of "one query per page" too literal. This idea is in fact a best practice and a good ideal goal (I won't waste words here on why).
However, the idea here is that we compose data fragments of the data we need in "one go". Pagination doesn't count as "one go".

Let's simplify your example a little, so I can talk more generally about it:

```
{
  field
  item {
    list(page: $page) {
      ...ListItem
    }
  }
}
```

This might represent a query of a page.
Suppose that this page displays some kind of data and details on a single item.

The list is displayed on the page and we get our first page displayed as well.

If we now fetch the second page, it's likely that we'd want to do this separately.

```
{
  item {
    list(page: $page) {
      ...ListItem
    }
  }
```

The easiest to do this would be to have a page component that can fragmentise our list. Subsequent pages to the first one would use this query rather than the "page query"
https://formidable.com/open-source/urql/docs/basics/ui-patterns/#infinite-scrolling

The same goes for sub lists.

```
{
  field
  item {
    list(page: $page) {
      ...ListItem
      subitemList {
        ... SubListItem
      }
    }
  }
}

{
  subitem {
    list(page: $page) {
      ... SubListItem
    }
  }
}
```

In other words, we are wishing to do two things here:
- Have our schema be designed in such a way that we can access a nested connection directly (because our UI needs it separately)
- Query a new separate page in a separate query for that page

That's at least how I would do it.

This of course doesn't need to be done with components or fragments, and could also be done by merging query results into local state, but the principle is the same.

**Reply by xoldd** [reply]
Thanks. You can mark this issue as answered btw.
